sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/DeploymentStep.java

```diff
@@ -176,25 +177,48 @@ private void setTaskStatus(Protos.TaskID taskID, Status status) {
         }
     }
 
-    private Status getStatus(Map<Protos.TaskID, TaskStatusPair> tasks) {
+    @VisibleForTesting
+    Status getStatus(Map<Protos.TaskID, TaskStatusPair> tasks) {
         if (tasks.isEmpty()) {
             return Status.PENDING;
         }
 
+        Set<Status> statuses = new HashSet<>();
         for (Map.Entry<Protos.TaskID, TaskStatusPair> entry : tasks.entrySet()) {
             String taskId = entry.getKey().getValue();
             Status status = entry.getValue().getStatus();
             logger.info("TaskId: {} has status: {}", taskId, status);
-            if (!status.equals(Status.COMPLETE)) {
-                // Keep and log current status
-                return super.getStatus();
-            }
+
+            statuses.add(status);
+        }
+
+        // A DeploymentStep should have the "least" status of its consituent tasks.
+        // We can rely on the order of Status as returned by its getOrder method.
+        // For example:
+        // 1 PENDING, 2 STARTING => PENDING
+        // 2 STARTING, 1 COMPLETE => STARTING
+        // 3 COMPLETE => COMPLETE
+        if (statuses.contains(Status.ERROR)) {
+            return Status.ERROR;
+        } else if (statuses.contains(Status.PENDING)) {
+            return Status.PENDING;
+        } else if (statuses.contains(Status.PREPARED)) {
+            return Status.PREPARED;
+        } else if (statuses.contains(Status.STARTING)) {
+            return Status.STARTING;
+        } else if (statuses.contains(Status.COMPLETE) && statuses.size() == 1) {
+            return Status.COMPLETE;
         }
 
-        return Status.COMPLETE;
+        // If we don't explicitly handle the new status,
+        // we will simply return the previous status.
+        logger.warn("The minimum status of the set of task statuses, {}, is not explicitly handled. " +
```

Comment:
This be noisy in the case where `contains(COMPLETE) && statuses.size() != 1`. I assume this code shouldn't be reachable in practice?

Maybe add an explicit case for that above? Something like:
```
} else if (statuses.contains(Status.COMPLETE)) {
    if (statuses.size() == 1) {
        return Status.COMPLETE;
    } else {
        // ?
    }
}
```

Reply:
It should not be reachable in practice.

The else I would put is just what's outside the the if/else if and then PMD will yell at me that I can collapse ifs.